File: dataset/human/c/file_9702.c
#include <stdio.h>
#define N 100000
#define NIL -1

typedef struct{int p, l, r;}Tree;

Tree t[N];
int n, d[N], k[N];

int Depth(int x, int p)
{
  d[x] = p;
  if(t[x].r != NIL) Depth(t[x].r, p);
  if(t[x].l != NIL) Depth(t[x].l, p+1);
}

void print(int x)
{
  int i, j, c;
  printf("node %d: parent = %d, depth = %d, ", x, t[x].p, d[x]);

  if(t[x].p == NIL) printf("root, ");
  else if(t[x].l == NIL) printf("leaf, ");
  else printf("internal node, ");

  printf("[");
  for(i = 0, c = t[x].l; c != NIL; i++, c = t[c].r){
    if(i > 0) printf(", ");
    printf("%d", c);
  }
  printf("]\n");
}

int main()
{
  int i, j, c, id, d, l, r;

  scanf("%d", &n);
  for(i = 0; i < n; i++){
    t[i].p = NIL;
    t[i].l = NIL;
    t[i].r = NIL;
  }
  
  for(i = 0; i < n; i++){
    scanf("%d%d", &id, &k[i]);
    for(j = 0; j < k[i]; j++){
      scanf("%d", &c);
      if(j == 0) t[id].l = c;
      else t[l].r = c;
      l = c;
      t[c].p = id;
    }
  }

  for(i = 0; i < n; i++){
    if(t[i].p == NIL) r =i;
  }
  Depth(r, 0);
  for(i = 0; i < n; i++) print(i);

  return 0;
}

File: dataset/human/java/file_9702.java
//API???????????????
import java.util.Scanner;
import java.io.IOException;
class Main {
  	public static void main(String[] args) throws IOException{
		Scanner scan = new Scanner(System.in);
		int n = scan.nextInt();

		Nodes nodes = new Nodes();
		Node[] node = new Node[n];
		nodes.nodes = node;
		for(int i = 0;i < n;i++){
			node[i] = new Node(i);
		}
		
		for(int i = 0;i < n; i++){
			int id = scan.nextInt();
			int left = scan.nextInt();
			int right = scan.nextInt();
			if(left != -1) node[left].parent = id;
			if(right != -1) node[right].parent = id;
			node[id].left = left;
			node[id].right = right;
			if(left != -1 && right != -1){
				node[left].sibling = right;
				node[right].sibling = left;
			}
		}
		
		int root = 0;
		
		while(node[root].parent != -1){
			root = node[root].parent;
		}
		
		nodes.setDepth(root, 0);
		node[root].height = nodes.setHeight(root);
		
		for(int i = 0; i < n; i++){
			StringBuilder sb = new StringBuilder();
			int d = 0;
			if(node[i].left != -1) d++;
			if(node[i].right != -1) d++;
			
			sb.append("node " + i + ": parent = " + node[i].parent + ", sibling = " + node[i].sibling + ", degree = " + d + ", depth = " + node[i].depth + ", height = " + node[i].height + ", ");
			//????????????????????°
			if(node[i].parent == -1) sb.append("root\r\n");
			//???????????????????????????????????°
			else if(d == 0) sb.append("leaf\r\n");
			//????????\???
			else sb.append("internal node\r\n");
			System.out.print(sb);
		}
	}
}

class Node{
	int id;
	int parent = -1;
	int left = -1;
	int right = -1;
	int sibling = -1;
	int depth = -1;
	int height = -1;

	public Node(int id) {
		this.id = id;
	}
}

class Nodes{
	/*
	List<Node> nodes;
	/*/
	Node[] nodes;
	/**/
	
	//-1??????????????§??±????????£?????????
	public void setDepth(int id, int d){
		int l = nodes[id].left;
		int r = nodes[id].right;
		if(l != -1) setDepth(l, d + 1);
		if(r != -1) setDepth(r, d + 1);
		nodes[id].depth = d;
	}
	
	public int setHeight(int id){
		if(id == -1) return -1;
		
		int l = nodes[id].left;
		int r = nodes[id].right;
		nodes[id].height = Math.max(setHeight(l), setHeight(r)) + 1; 
		return nodes[id].height;
	}
}
